Makefile

```diff
@@ -27,7 +36,22 @@ test:
 ## Release Management ##
 ########################
 
+GIT_REPO_AVAILABLE := $(shell git rev-parse --is-inside-work-tree 2>/dev/null)
+
+# Generate a unique tag based solely on the git hash.
+# This will be the identifier used for deployment via terraform.
+ifdef GIT_REPO_AVAILABLE
+IMAGE_TAG := $(shell git rev-parse HEAD)
+else
+IMAGE_TAG := "uknown-dev.$(DATE)"
+endif
+
 release-build:
```

Comment:
Good point. Couple of options:

* Option 1 - application template repos also have their own `release-build` command (could use Make, but doesn't have to) that is called as part of the application's ci-app.yml. The application's version of release-build doesn't have to tag the release, since the template-infra version will do that:
  * Cons: build command in two places, and while 99% of the build logic is within Dockerfile and code, there's still a small chance that difference in build command line arguments could produce a different build in CI than what is used for release
* Option 2 - we can run `release-build` as part of template-infra's ci-infra.yml, so we still get CI test coverage of build process
  * Cons: things like tests and linting in ci-app.yml can't use the docker image to run the tests, which potentially means CI and production are using slightly different environments

I believe what you suggested is Option 1?

Thoughts on these, or any other options you can think of?

Reply:
Or Option 3 could be something like
- template-infra Makefile's release-build calls template-application Makefile's release-build, and tags the image afterwards. For example:

```
# template-infra/Makefile

release-build:
  IMAGE_NAME := $(PROJECT_NAME)-$(APP_NAME)
  IMAGE_NAME=$(IMAGE_NAME) $(MAKE) -C $(APP_NAME) release-build
  docker tag $(IMAGE_NAME):latest $(IMAGE_NAME):$(IMAGE_TAG)
```

```
# template-application/Makefile

release-build:
  docker build --tag $(IMAGE_NAME):latest --target release .
```

Could Makefile commands have return values or arguments? If so we can make it even more explicit by reading the return value or something.

Reply:
> Option 2 - we can run release-build as part of template-infra's ci-infra.yml, so we still get CI test coverage of build process
>
> Cons: things like tests and linting in ci-app.yml can't use the docker image to run the tests, which potentially means CI and production are using slightly different environments

I think the con here is okay? The release build should be minimal, so it wouldn't have tests, developer/test tools, etc. anyways.

There is another con in that each application might want a "healthcheck" / "ping" as part of the CI to verify that it actually starts successfully (not just that it builds), but that's a bit of a rabbit hole / out-of-scope.

I think this option is reasonable since `release-build` starts to move into the deployment flow. But we'd probably need a _separate_ workflow so it gets triggered when the app files change instead of when the infra files change 😵

Reply:
Option 3 seems okay too. I might be tempted into this one just to avoid the confusion of a `template-infra` CI workflow that actually depends on application files changing.

Reply:
I pushed up a version with the last option (just passing in OPTS). Not sure how to pass in `TAGS` and make it work without TAGS since passing in an empty arg to `--tag` like `docker build --tag --target release .` will fail.

I tested from top level Makefile
<img width="1345" alt="image" src="https://user-images.githubusercontent.com/447859/193337620-43fc0680-6bb9-47cb-9fb5-38e9b2ac49fe.png">

and also from the app level Makefile
<img width="868" alt="image" src="https://user-images.githubusercontent.com/447859/193337709-e62ecc0c-d79b-4be2-9f39-a6bb2abe4875.png">

if you think it looks good i'll merge, and update the #106 CI/CD proposal

Reply:
yep that looks good to me, thanks!